ocaml/xapi/certificates_sync.ml

```diff
@@ -0,0 +1,144 @@
+module G = Gencertlib
+module GP = Gencertlib.Pem
+module GV = Gencertlib.Validation
+open Api_errors
+open Rresult
+
+let ( let* ) = ( >>= )
+
+let is_tls_verification_enabled () =
+  Localdb.get_with_default Constants.tls_verification_enabled "false"
+  |> bool_of_string
+
+let is_self_signed cert =
+  R.return
+    X509.(
+      Distinguished_name.equal (Certificate.subject cert)
+        (Certificate.issuer cert))
+
+let pool_size ~__context = Db.Pool.get_all ~__context |> List.length
+
+let remove_cert ~__context cert =
+  info "Removing certificate %s" (Ref.string_of cert) ;
+  Db.Certificate.destroy ~__context ~self:cert ;
+  (* the host cert is under a fixed name - so the new cert in the file
+     system already replaced the old one. We must not remove it from the
+     file system *)
+  Ok ()
+
+(** verify and install a certificate or return with an error *)
+let install ~__context ~host cert =
+  try
+    let host_pem = cert.GP.host_cert in
+    let* host_cert =
+      Cstruct.of_string host_pem
+      |> X509.Certificate.decode_pem
+      |> R.reword_error (fun (`Msg msg) ->
+             D.info {|Failed to decode certificate because "%s"|} msg ;
+             `Msg (server_certificate_invalid, []))
+    in
+    let* self_signed = is_self_signed host_cert in
+    match self_signed with
+    | true ->
+        let raw = X509.Certificate.encode_pem host_cert in
+        let pem = Cstruct.to_string raw in
+        (* pass own cert as root *)
+        let ref =
+          Xapi_host.add_certificate_to_db ~__context ~host host_cert [raw]
+        in
+        ( match pool_size ~__context with
+        | 1 ->
+            let uuid = Db.Certificate.get_uuid ~__context ~self:ref in
+            Certificates.(
+              pool_install CA_Certificate ~__context ~name:uuid ~cert:pem)
+        | _ ->
+            debug "Not installing updated self-signed host cert in pool"
+        ) ;
+        R.ok ()
+    | false -> (
+        let ca_cert_bundle_path = !Xapi_globs.stunnel_cert_path in
+        let pem_leaf = cert.GP.host_cert in
+        let pkcs8_private_key = cert.GP.private_key in
+        let pem_intermediates = String.concat "\n\n" cert.GP.other_certs in
+        let time = Ptime_clock.now () in
+        let tls_on = is_tls_verification_enabled () in
+        match
+          GV.validating_trust_anchors ~pem_leaf ~pkcs8_private_key
+            ~pem_intermediates ~time ~ca_cert_bundle_path
+        with
+        | Error _ as error_msg when tls_on ->
+            error "Host certificate is invalid but required by TLS" ;
+            error_msg
+        | Ok (host_cert, roots) ->
+            let hashes = List.map X509.Certificate.encode_pem roots in
+            let _ref =
+              Xapi_host.add_certificate_to_db ~__context ~host host_cert hashes
+            in
+            R.ok ()
+        | Error _ as error_msg ->
+            warn "The host certificate is invalid but TLS is off" ;
+            error_msg
```

Comment:
Won't returning the error as-is make the host enter maintenance mode, even if the certificates is not verified?

If a host was just updated and it fails here it will look like the host failed to upgrade. I think logging this and continuing with an `Ok ()`

Installing an invalid certificate with this new version will just not be possible since we enforce the validation of the certificate.